<table class="m01-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/yy/dviz-course/blob/master/docs/m11-highdim/lab11supp-calc_eigenvector.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a href="https://yyahn.com/dviz-course/m11-highdim/lab11supp-calc_eigenvector/"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on Github</a>
  </td>
  <td>
    <a href="https://raw.githubusercontent.com/yy/dviz-course/master/docs/m11-highdim/lab11supp-calc_eigenvector.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View raw on Github</a>
  </td>
</table>

# Module 11: Eigenvalue and eigenvector calculation

In [ ]:
import numpy as np

In [2]:
# Enter the data
mat = np.asmatrix([(0,2,1,2,-1),
                   (2,6,4,5,-1)])

In [3]:
mat

matrix([[ 0,  2,  1,  2, -1],
        [ 2,  6,  4,  5, -1]])

In [4]:
# Compute the covariance matrix
cov = np.cov(mat)

In [5]:
cov

array([[1.7 , 3.55],
       [3.55, 7.7 ]])

In [6]:
# Compute the eigenvectors and eigenvalues
eigvals, eigvecs = np.linalg.eig(cov)

In [7]:
eigvals

array([0.05215104, 9.34784896])

The 2nd eigenvalue is larger, so this is the *first* principal component.

In [8]:
eigvecs

array([[-0.90704462, -0.42103452],
       [ 0.42103452, -0.90704462]])

The 2nd column is the 2nd eigenvector: [ -0.42103452, -0.90704462]. This corresponds to the 2nd eigenvalue. So the first component of this vector is -0.42103452.

Let's check!

In [9]:
eigvecs[:,1]

array([-0.42103452, -0.90704462])

If we multiply this vector with the covariance matrix, we should get the vector times the corresponding eigenvalue because 

$$ A \vec{v} = \lambda \vec{v}$$

when $\vec{v}$ is a eigenvector and $\lambda$ is the corresponding eigenvalue. 

In [10]:
cov.dot(eigvecs[:,1])

array([-3.93576707, -8.47891608])

In [11]:
eigvals[1] * eigvecs[:,1]

array([-3.93576707, -8.47891608])

Indeed! 